test_helpers/main.go

```diff
@@ -0,0 +1,166 @@
+}
+
+// TarantoolInstance is a data for instance graceful shutdown and cleanup.
+type TarantoolInstance struct {
+	// Cmd is a Tarantool command. Used to kill Tarantool process.
+	Cmd *exec.Cmd
+
+	// WorkDir is a directory with tarantool data. Cleaned up after run.
+	WorkDir string
+}
+
+func isReady(server string, opts *tarantool.Opts) error {
+	var err error
+	var conn *tarantool.Connection
+	var resp *tarantool.Response
+
+	conn, err = tarantool.Connect(server, *opts)
+	if err != nil {
+		return err
+	}
+	if conn == nil {
+		return errors.New("Conn is nil after connect")
+	}
+	defer conn.Close()
+
+	resp, err = conn.Ping()
+	if err != nil {
+		return err
+	}
+	if resp == nil {
+		return errors.New("Response is nil after ping")
+	}
+
+	return nil
+}
+
+// StartTarantool starts a tarantool instance for tests
+// with specifies parameters (refer to StartOpts).
+// Process must be stopped with StopTarantool.
+func StartTarantool(startOpts StartOpts) (TarantoolInstance, error) {
+	// Prepare tarantool command.
+	var inst TarantoolInstance
+	inst.Cmd = exec.Command("tarantool", startOpts.InitScript)
+
+	inst.Cmd.Env = append(
+		os.Environ(),
+		fmt.Sprintf("TEST_TNT_WORK_DIR=%s", startOpts.WorkDir),
+		fmt.Sprintf("TEST_TNT_LISTEN=%s", startOpts.Listen),
+	)
+
+	// Clean up existing work_dir.
+	err := os.RemoveAll(startOpts.WorkDir)
+	if err != nil {
+		return inst, err
+	}
+
+	// Create work_dir.
+	err = os.Mkdir(startOpts.WorkDir, 0755)
+	if err != nil {
+		return inst, err
+	}
+
+	inst.WorkDir = startOpts.WorkDir
+
+	// Start tarantool.
+	err = inst.Cmd.Start()
+	if err != nil {
+		return inst, err
+	}
+
+	// Try to connect and ping tarantool.
+	// Using reconnect opts do not help on Connect,
+	// see https://github.com/tarantool/go-tarantool/issues/136
+	time.Sleep(startOpts.WaitStart)
+
+	opts := tarantool.Opts{
+		Timeout:    500 * time.Millisecond,
+		User:       startOpts.User,
+		Pass:       startOpts.Pass,
+		SkipSchema: true,
+	}
+
+	var i uint
+	for i = 0; i <= startOpts.ConnectRetry; i++ {
+		err = isReady(startOpts.Listen, &opts)
+
+		// Both connect and ping is ok.
+		if err == nil {
+			break
+		}
+
+		if i != startOpts.ConnectRetry {
+			time.Sleep(startOpts.RetryTimeout)
+		}
+	}
+
+	return inst, err
+}
+
+// StopTarantool stops a tarantool instance started
+// with StartTarantool. Waits until any resources
+// associated with the process is released.
+// Cleans work directory after stop. If something went wrong, fails.
+func StopTarantool(inst TarantoolInstance) {
+	if inst.Cmd != nil && inst.Cmd.Process != nil {
+		if err := inst.Cmd.Process.Kill(); err != nil {
+			log.Fatalf("Failed to kill tarantool (pid %d), got %s", inst.Cmd.Process.Pid, err)
+		}
+
+		// Wait releases any resources associated with the Process.
+		if _, err := inst.Cmd.Process.Wait(); err != nil {
+			log.Fatalf("Failed to wait for Tarantool process to exit, got %s", err)
+		}
+
+		inst.Cmd = nil
+	}
+
+	if inst.WorkDir != "" {
```

Comment:
It doesn't seem intuitive to me that `StopTarantool` do a cleanup.
I try to use https://pkg.go.dev/testing#B.Cleanup to cleanup test artifacts.
I suggest 3 variants to you:
* Leave as is
* Add boolean arguments to a function (like `isCleanup`)
* Create a function like `CleanupTarantoolStuff` and call it from the callback

Up to you.

Reply:
Reworked